Fix record history export

jasper-lyons opened the pull request.

jasper-lyons committed:
fix: always escape keys in exported objects

Previously we were erroneously testing to see if a value was a valid ruby key. This returned true even when the value was not a valid ruby key e.g. 

```
'A's Fake Name': {
```

```
Ghtc_Hyuik_1[1]_160x160: {
```
I've change this behavior so that all keys are wrapped in single quotes and, if any single quotes exist in the key string they are escaped e.g.

```
'A\'s Fake Name': {
```

```
'Ghtc_Hyuik_1[1]_160x160': {
```

jasper-lyons committed:
fix: escape escape character so that it's rendered

jasper-lyons committed:
Merge branch 'primeroIMS:master' into fix-record-history-export

aespinoza-quoin commented on migration/v1_to_v2/data/exporters/data_exporter.rb:
```diff
@@ -143,14 +143,8 @@ def value_to_ruby_string(value, include_blank = false)
-
   def key_to_ruby(key)
-    valid_key?(key) ? key : "'#{key}'"
+    "'#{ key.gsub(/'/, "\\'") }'"
```
If you receive an integer you will get an exception. so you can keep the validation and add your escape to those that are not valid:

```
def key_to_ruby(key)
  return key if valid_key?(key)

  "'#{ key to_s.gsub(/'/, "\\'") }'"
end
```

jasper-lyons replied:
I wouldn't want to rely on any `valid_key?` method for this, everything we're being passed should be a valid key given that it comes from an existing ruby `key: value` pair afaik.

I would instead just coerce everything to string before applying the `.gsub`.

jasper-lyons committed:
Forcing a light GC every 100 attachements imported

On some installations we can see upwards of 5GB of attachment data. When we try and load all of this into rails at once, depending on the available machine memory, we can see OOM exceptions which can bring everything down! Yay.

We're injecting a manual GC every 100 attachment loads which will drastically reduce the lifetime of the attachment objects.

jasper-lyons committed:
fix: nont string keys throwing errors

coerce all potential keys into strings before processing them

jasper-lyons committed:
fix: convert array of hashes to array of pairs

jasper-lyons committed:
Fix filename issues when exporting attachements

jasper-lyons committed:
refactor: use the procedurale approach, it's cleaner

jasper-lyons committed:
fix: use much smaller batch sizes for attachment exports

jasper-lyons committed:
fix: stop overiding default batch sizes

jasper-lyons committed:
Merge branch 'reduce-attachment-memory-consumption' into fix-record-history-export

aespinoza-quoin commented on migration/v1_to_v2/data/exporters/data_exporter.rb:
```diff
@@ -143,14 +143,8 @@ def value_to_ruby_string(value, include_blank = false)
-
   def key_to_ruby(key)
-    valid_key?(key) ? key : "'#{key}'"
+    "'#{ key.to_s.gsub(/'/, "\\'") }'"
```
we want to keep `valid_key?` method call here, in order to avoid unnecesary conversion of a valid key

aespinoza-quoin replied:
This method is also changed in this PR https://github.com/primeroIMS/primero-v2-migration/pull/4/files

aespinoza-quoin commented on migration/v1_to_v2/data/export_data.rb:
```diff
@@ -18,7 +18,7 @@ def exporters
 export_dir = "record-data-files-#{timestamp}"
 exporters.each do |exporter|
-  data_exporter = Object.const_get(exporter).new(export_dir: export_dir, batch_size: 250)
+  data_exporter = Object.const_get(exporter).new(export_dir: export_dir)
```
Keep 250 batch_size and overwrite in attachment_exporter

aespinoza-quoin replied:
[this change is not related with record history export, please create a new pr]

aespinoza-quoin commented on migration/v1_to_v2/data/exporters/attachment_exporter.rb:
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require_relative('data_exporter.rb')
+require 'erb'
```
[these changes is not related with record history export, please create a new pr]

aespinoza-quoin commented on migration/v1_to_v2/data/exporters/attachment_exporter.rb:
```diff
@@ -152,34 +158,53 @@ def add_date_to_file_for_attachment(date)
+     initialize_script_for_attachment(folder_to_save, type, sufix)
+
+     attachments = []
+     data_object_names.each do |type|
```
this is a reduce/inject to avoid initialize array, please create a new method

aespinoza-quoin commented on migration/v1_to_v2/data/exporters/attachment_exporter.rb:
```diff
@@ -152,34 +158,53 @@ def add_date_to_file_for_attachment(date)
+       end
+
+       @output.puts "# force ruby to gargabe collect here as attachmented files can build up in memory!"
+       @output.puts "GC.start"
```
GC will be enable for every iteration. is not need to stop manually?

pnabutovsky commented:
We are holding off on merging this. The main performance motive for this PR has been addressed elsewhere.
